File: src/infrastructure/validators/process-number.ts
import { IValidationResult, ok, fail } from './result';

const STRICT = /^\d{5}\.\d{6}\/\d{2}-\d{2}$/;
// Tolerante: 17 dígitos no total, com separadores variados.
const LOOSE = /(\d{5})[.\s]*(\d{6})[/\s]*(\d{2})[-\s]*(\d{2})/;

/** Valida/normaliza o número de processo para `00000.000000/00-00`. */
export function validateProcessNumber(raw: string | null | undefined): IValidationResult {
  const value = (raw ?? '').trim();
  if (!value) return fail('Número de processo vazio');

  if (STRICT.test(value)) return ok(value);

  const m = LOOSE.exec(value);
  if (!m) return fail('Formato de número de processo não reconhecido');

  const normalized = `${m[1]}.${m[2]}/${m[3]}-${m[4]}`;
  return { valid: true, normalized, messages: [`Número de processo normalizado: "${value}"`] };
}

File: app/validators/process_number.py
"""Validação do número de processo administrativo.

Formato de referência: 21050.003108/99-98
Estrutura: 5 dígitos . 6 dígitos / 2 dígitos - 2 dígitos
"""
from __future__ import annotations

import re

from .base import ValidationResult

_STRICT = re.compile(r"^\d{5}\.\d{6}/\d{2}-\d{2}$")
# Captura tolerante a ruído de OCR entre os blocos, mas só com separadores
# plausíveis (espaço, ponto, barra, hífen) — evita aceitar lixo como '@' '*'.
_LOOSE = re.compile(r"(\d{5})[\s./-]{0,3}(\d{6})[\s./-]{0,3}(\d{2})[\s./-]{0,3}(\d{2})")


def format_process(a: str, b: str, c: str, d: str) -> str:
    return f"{a}.{b}/{c}-{d}"


def is_valid(raw: str) -> bool:
    return validate(raw).valid


def validate(raw: str) -> ValidationResult:
    text = (raw or "").strip()
    if not text:
        return ValidationResult(False, None, ["Número de processo vazio"])

    if _STRICT.match(text):
        return ValidationResult(True, text, [])

    match = _LOOSE.search(text)
    if not match:
        return ValidationResult(
            False, None, ["Número de processo fora do padrão (00000.000000/00-00)"]
        )

    normalized = format_process(*match.groups())
    messages: list[str] = []
    if normalized != text:
        messages.append("Processo normalizado a partir de formato divergente")
    return ValidationResult(True, normalized, messages)
